### PR TITLE
Add handler for refreshing invalid token

### PR DESCRIPTION
**Pull request by trapacska**
Untitled Diagram (13)

### CODE REVIEW

File: client/client_with_secret_test.go [hunk 1]
```diff
@@ -215,3 +217,67 @@ func Test_GivenTokenSourceWithTokenThatWillNotExpireBetweenRequests_WhenTokenSto
...
+	assert.NoError(t, err)
+	assert.Equal(t, http.StatusUnauthorized, resp.StatusCode)
+
+	mockedAuthService.AssertExpectations(t)
+	mockedClient.AssertExpectations(t)
+	mockedAuthService.AssertNotCalled(t, "Token", "refresh_token")
+	mockedClient.AssertNotCalled(t, "Test", "refreshed-access-token")
+}
+
+func Test_GivenAServerThatDoesnotRejectHTTPCall_WhenAGetCallIsFired_ThenExpectTheClientToDoNotRenewAccessToken(t *testing.T) {
```
Comment (trapacska) [inline client/client_with_secret_test.go:253]: This two tests could go to a table test

File: client/invalid_token_refresher_transport.go [hunk 2]
```diff
@@ -0,0 +1,28 @@
...
+	tokenSrc oauth2.TokenSource
+}
+
+func (t *invalidTokenRefresherTransport) RoundTrip(req *http.Request) (*http.Response, error) {
+	resp, err := t.base.RoundTrip(req)
+	if err != nil {
+		return resp, err
+	}
+
+	if resp.StatusCode == http.StatusUnauthorized {
```
Comment (zsolt-marta-bitrise) [inline client/invalid_token_refresher_transport.go:20]: What happens if this unauthorized response is not solved by re-requesting the token? It will get into an infinite cycle, right?
This might be a too broad condition

File: client/client_with_secret_test.go [hunk 3]
```diff
@@ -215,3 +217,67 @@ func Test_GivenTokenSourceWithTokenThatWillNotExpireBetweenRequests_WhenTokenSto
...
+		client.WithBaseURL(ts.URL)).HTTPClient()
+
+	// Then
+	resp, err := c.Get(ts.URL)
+	assert.NoError(t, err)
+	assert.Equal(t, http.StatusUnauthorized, resp.StatusCode)
+
+	mockedAuthService.AssertExpectations(t)
+	mockedClient.AssertExpectations(t)
+	mockedAuthService.AssertNotCalled(t, "Token", "refresh_token")
```
Comment (zsolt-marta-bitrise) [inline client/client_with_secret_test.go:249]: I'm confused, shouldn't we assert that the token is refreshed?

File: client/invalid_token_refresher_transport.go (hunk 2 above)
Comment (zsolt-marta-bitrise) [reply client/invalid_token_refresher_transport.go:20]: Oh, I just realized that `t.base.RoundTrip(req)` will prevent that

File: client/client_with_secret.go [hunk 4]
```diff
@@ -89,16 +89,26 @@ func (cws *WithSecret) HTTPClient(opts ...HTTPClientOption) *http.Client {
 		opt(clientOpts)
 	}
 
-	if clientOpts.baseClient == nil {
-		return creds.Client(clientOpts.context)
+	c := &http.Client{}
```
Comment (zsolt-marta-bitrise) [inline client/client_with_secret.go:92]: Can you please call this `client`, or something more verbose?

File: client/client_with_secret_test.go [hunk 5]
```diff
@@ -215,3 +217,67 @@ func Test_GivenTokenSourceWithTokenThatWillNotExpireBetweenRequests_WhenTokenSto
...
+	// When
+	c := client.NewWithSecret("my-client-id", "my-secret",
+		client.WithBaseURL(ts.URL)).HTTPClient()
+
+	// Then
+	resp, err := c.Get(ts.URL)
+	assert.NoError(t, err)
+	assert.Equal(t, http.StatusUnauthorized, resp.StatusCode)
+
+	mockedAuthService.AssertExpectations(t)
```
Comment (adborbas) [inline client/client_with_secret_test.go:247]: We have expectation for `mockedAuthService` both as `.AssertExpectations(t)` and `.AssertNotCalled(t....` which is a bit hard to follow. Could we assert only one way? Ideally not using the `AssertExpectations`.

File: client/client_with_secret_test.go (hunk 3 above)
Comment (trapacska) [reply client/client_with_secret_test.go:249]: No, as we are not using refresh_token from OIDC protocol. In oauth2 s2c the refresh means "request a new one"

File: client/client_with_secret_test.go [hunk 6]
```diff
@@ -215,3 +217,43 @@ func Test_GivenTokenSourceWithTokenThatWillNotExpireBetweenRequests_WhenTokenSto
...
+		mockedClient.
+			On("Test", "initial-access-token").Return().
+			Times(testCase.expectedNoOfCalls)
+
+		// When
+		c := client.NewWithSecret("my-client-id", "my-secret",
+			client.WithBaseURL(ts.URL)).HTTPClient()
+
+		// Then
+		resp, err := c.Get(ts.URL)
```
Comment (adborbas) [inline client/client_with_secret_test.go:252]: Just two small things
- shouldn't `resp, err := c.Get(ts.URL)` be in the `When`
- and`c := client.NewWithSecret...` in the `Given` part?

**Review by tadam313 — DISMISSED** [review]
@trapacska awesome, and elegant solution! well done 🙇

**Review by zsolt-marta-bitrise — APPROVED** [review]
